File: wpf_prototype/common/Controller/JoystickSerialiser.cs
namespace Controller;

public static class JoystickSerialiser
{
    private const int FloatSize = sizeof(float);
    private const int BytesRequired = (2 * FloatSize);

    private const int XPosition = 0;
    private const int YPosition = XPosition + FloatSize;

    public static (float x, float y) DeserialiseJoystickState(ReadOnlySpan<byte> data)
    {
        return (
            BitConverter.ToSingle(data[..YPosition]),
            BitConverter.ToSingle(data[YPosition..])
        );
    }

    public static byte[] SerialiseJoystickState((float x, float y) joystickPosition)
    {
        var result = new byte[BytesRequired];
        Array.Copy(BitConverter.GetBytes(joystickPosition.x), 0, result, XPosition, FloatSize);
        Array.Copy(BitConverter.GetBytes(joystickPosition.y), 0, result, YPosition, FloatSize);
        return result;
    }
}

File: prototype/android_viewer/PipelineControllerViewer/app/src/main/java/com/danjbower/pipelinecontrollerviewer/data/JoystickSerialiser.kt
package com.danjbower.pipelinecontrollerviewer.data

import java.nio.ByteBuffer
import java.nio.ByteOrder

object JoystickSerialiser
{
    private const val FLOAT_SIZE = 4
    private const val BYTES_REQUIRED = 2 * FLOAT_SIZE

    private const val X_POSITION = 0
    private const val Y_POSITION = X_POSITION + FLOAT_SIZE

    fun deserialiseJoystickState(data: ByteArray): Pair<Float, Float> {
        require(data.size >= BYTES_REQUIRED) {
            "Not enough bytes to deserialize joystick state (need $BYTES_REQUIRED, got ${data.size})"
        }
        val x = ByteBuffer.wrap(data, X_POSITION, FLOAT_SIZE).order(ByteOrder.LITTLE_ENDIAN).float
        val y = ByteBuffer.wrap(data, Y_POSITION, FLOAT_SIZE).order(ByteOrder.LITTLE_ENDIAN).float
        return x to y
    }

    fun serialiseJoystickState(joystickPosition: Pair<Float, Float>): ByteArray {
        val (x, y) = joystickPosition
        val result = ByteArray(BYTES_REQUIRED)
        writeFloat(result, X_POSITION, x)
        writeFloat(result, Y_POSITION, y)
        return result
    }

    private fun writeFloat(dest: ByteArray, offset: Int, value: Float) {
        val buf = ByteBuffer.allocate(FLOAT_SIZE).order(ByteOrder.LITTLE_ENDIAN)
        buf.putFloat(value)
        System.arraycopy(buf.array(), 0, dest, offset, FLOAT_SIZE)
    }
}
